File: test/corpus/propagation/mut_ref_vec_push_propagation.exs
defmodule RustQ.Corpus.Propagation.MutRefVecPushPropagation do
  @moduledoc "Propagate fallible call arguments passed to push on a mutable Vec reference."

  use RustQ.Meta

  alias RustQ.Type, as: R

  @spec make_term() :: term()
  defrust(make_term(), do: 0)

  @spec decode_value() :: R.nif_result(term())
  defrust(decode_value(), do: {:ok, make_term()})

  @spec run(R.mut_ref(R.vec(term()))) :: R.nif_result(R.unit())
  defrust run(values) do
    values.push(decode_value())
    :ok
  end
end

File: test/corpus/propagation/mut_ref_vec_push_propagation.rs
fn make_term<'a>() -> Term<'a> {
    0
}

fn decode_value<'a>() -> NifResult<Term<'a>> {
    Ok(make_term())
}

fn run<'a>(values: &mut Vec<Term<'a>>) -> NifResult<()> {
    values.push(decode_value()?);
    Ok(())
}
